### PR TITLE
feat: add unique method on the array namespace

### PR DESCRIPTION
**Pull request by sgulseth**

### CODE REVIEW

File: test/functions.test.ts [hunk 1]
```diff
@@ -29,4 +29,63 @@ t.test('Functions', async (t) => {
       t.same(data.deep.time, data.topTime)
     })
   })
+
+  t.test('Array namespace', async (t) => {
```
Comment (atombender) [inline test/functions.test.ts:33]: Isn't the function already covered by the GROQ test suite?

**Review by atombender — CHANGES_REQUESTED** [review]

File: src/evaluator/functions.ts [hunk 2]
```diff
@@ -441,11 +441,44 @@ diff.changedOnly = () => {
...
+      case 'boolean':
+      case 'datetime':
+        if (result.includes(element) === false) {
+          result.push(element)
+        }
+        break
+      default:
+        result.push(element)
+    }
+  }
```
Comment (atombender) [inline src/evaluator/functions.ts:471]: Avoid linear search?

```suggestion
  const result = []
  const added = new Set()
  for await (const iter of value) {
    const element = await iter.get()
    switch (iter.type) {
      case 'number':
      case 'string':
      case 'boolean':
      case 'datetime':
        if (!added.has(element.data)) {
          added.add(element.data)
          result.push(element)
        }
        break
      default:
        result.push(element)
    }
  }
```

Comment (sgulseth) [reply src/evaluator/functions.ts:471]: Perf! Good point

File: test/functions.test.ts (hunk 1 above)
Comment (sgulseth) [reply test/functions.test.ts:33]: Not atm, and I wanted something to test with locally. Can add some tests to the test suite though!

Comment (atombender) [reply test/functions.test.ts:33]: We have tests [here](https://github.com/sanity-io/groq-test-suite/pull/54).

Comment (sgulseth) [reply test/functions.test.ts]: Ah! Removed, tested and verified with `GROQTEST_SUITE=../groq-test-suite/suite.ndjson ./test/generate.sh` from the branch

**Comment by atombender** [comment]
Great. Awaiting review on [spec PR](https://github.com/sanity-io/GROQ/pull/82) from @judofyr.

File: src/evaluator/functions.ts [hunk 3]
```diff
@@ -456,11 +456,47 @@ diff.changedOnly = () => {
 }
 diff.changedOnly.arity = 3
 
+const array: FunctionSet = {}
+
+array.unique = async function (args, scope, execute) {
+  if (args.length > 1) {
+    throw new Error('Incorrect number of arguments to function "array::unique"')
+  }
```
Comment (judofyr) [inline src/evaluator/functions.ts:464]: This should be handled by setting `array.unique.arity = 1` instead. That will correctly treat `array::unique(foo, bar)` as a compile-time error.

File: src/evaluator/functions.ts [hunk 4]
```diff
@@ -456,11 +456,47 @@ diff.changedOnly = () => {
...
+          result.push(iter.data)
+        }
+        break
+      default:
+        const element = await iter.get()
+        result.push(element)
+    }
+  }
+
+  return fromJS(result)
```
Comment (judofyr) [inline src/evaluator/functions.ts:491]: It's possible to implement this in a streaming manner:

```suggestion
  return new new StreamValue(async function* () {
    const added = new Set()
    for await (const iter of value) {
      switch (iter.type) {
        case 'number':
        case 'string':
        case 'boolean':
        case 'datetime':
          if (!added.has(iter.data)) {
            added.add(iter.data)
            yield iter
          }
          break
        default:
          yield iter
      }
    }
  })
```

**Review by judofyr — CHANGES_REQUESTED** [review]

**Review by judofyr — APPROVED** [review]

**Pull request merged**